beacon-chain/forkchoice/protoarray/optimistic_sync.go

```diff
@@ -0,0 +1,85 @@
+package protoarray
+
+import (
+	"fmt"
+
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/config/params"
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+)
+
+// This returns the minimum and maximum slot of the synced_tips tree
+func (f *ForkChoice) boundarySyncedTips() (types.Slot, types.Slot) {
+	f.syncedTips.RLock()
+	defer f.syncedTips.RUnlock()
+
+	min := params.BeaconConfig().FarFutureSlot
+	max := types.Slot(0)
+	for _, slot := range f.syncedTips.validatedTips {
+		if slot > max {
+			max = slot
+		}
+		if slot < min {
+			min = slot
+		}
+	}
+	return min, max
+}
+
+// Optimistic returns true if this node is optimistically synced
+// WARNING: this function does not check if slot corresponds to the
+//          block with the given root. An incorrect response may be
+//          returned when requesting earlier than finalized epoch due
+//          to pruning of non-canonical branches. A requests for a
+//          combination root/slot of an available block is guaranteed
+//          to yield the correct result. The caller is responsible for
+//          checking the block's availability. A consensus bug could be
+//          a cause of getting this wrong, so think twice before passing
+//          a wrong pair.
+func (f *ForkChoice) Optimistic(root [32]byte, slot types.Slot) (bool, error) {
+
```

Comment:
Given that this is a recursive function, please add a context.Context and check for context error to exit early.
```suggestion
func (f *ForkChoice) Optimistic(ctx context.Context, root [32]byte, slot types.Slot) (bool, error) {
  if ctx.Err() != nil {
    return false, ctx.Err()
  }

```